packages/app-builder-lib/src/node-module-collector/packageManager.ts

```diff
@@ -0,0 +1,112 @@
+// copy from https://github.com/egoist/detect-package-manager/blob/main/src/index.ts
+// and merge https://github.com/egoist/detect-package-manager/pull/9 to support Monorepo
+import { promises as fs } from "fs"
+import { resolve, dirname } from "path"
+import { exec } from "child_process"
+import { promisify } from "util"
+
+const execa = promisify(exec)
+
+export type PM = "npm" | "yarn" | "pnpm" | "bun"
+
+/**
+ * Check if a path exists
+ */
+async function pathExists(p: string) {
+  try {
+    await fs.access(p)
+    return true
+  } catch {
+    return false
+  }
+}
+
+const cache = new Map()
+
+/**
+ * Check if a global pm is available
+ */
+function hasGlobalInstallation(pm: PM): Promise<boolean> {
+  const key = `has_global_${pm}`
+  if (cache.has(key)) {
+    return Promise.resolve(cache.get(key))
+  }
+  const execa = promisify(exec)
+
+  return execa(`${pm} --version`)
+    .then(res => {
+      return /^\d+.\d+.\d+$/.test(res.stdout)
+    })
+    .then(value => {
+      cache.set(key, value)
+      return value
+    })
+    .catch(() => false)
+}
+
+function getTypeofLockFile(cwd = "."): Promise<PM | null> {
+  const key = `lockfile_${cwd}`
+  if (cache.has(key)) {
+    return Promise.resolve(cache.get(key))
+  }
+
+  return Promise.all([
+    pathExists(resolve(cwd, "yarn.lock")),
+    pathExists(resolve(cwd, "package-lock.json")),
+    pathExists(resolve(cwd, "pnpm-lock.yaml")),
+    pathExists(resolve(cwd, "bun.lockb")),
+  ]).then(([isYarn, isNpm, isPnpm, isBun]) => {
+    let value: PM | null = null
+
+    if (isYarn) {
+      value = "yarn"
+    } else if (isPnpm) {
+      value = "pnpm"
+    } else if (isBun) {
+      value = "bun"
+    } else if (isNpm) {
+      value = "npm"
+    }
```

Comment:
Can we use this for default fallback of npm?
```
} else {
  value = "npm"
}
```
Then the function return type is always `Promise<PM>`

Reply:
Yes. We can return the default fallback pf npm.